sys/kern/kmem.c

```diff
@@ -79,16 +79,15 @@ void kva_unmap(vaddr_t ptr, size_t size) {
 
   kasan_mark_invalid((void *)ptr, size, KASAN_CODE_KMEM_FREED);
 
-  vaddr_t va = (vaddr_t)ptr;
-  vaddr_t end = va + size;
-  while (va < end) {
+  vaddr_t va = ptr;
+  while (va < ptr + size) {
     vm_page_t *pg = kva_find_page(va);
     assert(pg != NULL);
     va += pg->size * PAGESIZE;
     vm_page_free(pg);
   }
 
-  pmap_kremove((vaddr_t)ptr, size);
+  pmap_kremove(ptr, size);
```

Comment:
This is incorrect: someone can allocate and map the freed page before our call to `pmap_kremove()`, and our call to `pmap_kremove()` will unmap a page that someone else allocated and mapped. I think the call to `pmap_kremove()` should be before the loop.

Reply:
Correction, we won't unmap someone else's mapping, since that someone will map the page at a different virtual address... still, I think we should first unmap the pages, and then free them.

Reply:
Ok. So `vm_pagelist_free` is necessary then.